### PR TITLE
Ldap

### PR DESCRIPTION
**Pull request by fvacek**

### CODE REVIEW

File: libshvbroker/src/brokerapp.h [hunk 1]
```diff
@@ -77,6 +77,10 @@ class SHVBROKER_DECL_EXPORT BrokerApp : public QCoreApplication
 	rpc::WebSocketServer* webSocketServer();
 #endif
 
+#ifdef WITH_SHV_LDAP
+	void setGroupForLdapUser(const std::string_view& user_name, const std::string_view& group_name);
```
Comment (fvacek) [inline libshvbroker/src/brokerapp.h:81]: What about to save current user LDAP groups in ClientConnection. They can be sent by `Q_SIGNAL AuthThread::resultReady(login_result, user_groups)`

File: libshvbroker/src/brokerapp.cpp [hunk 2]
```diff
@@ -517,9 +523,70 @@ bool BrokerApp::checkTunnelSecret(const std::string &s)
...
+#ifdef WITH_SHV_LDAP
+		if (BrokerApp::instance()->cliOptions()->ldapHostname_isset()) {
+			try {
+				auto ldap = ldap::Ldap::create(BrokerApp::instance()->cliOptions()->ldapHostname());
+				ldap->setVersion(ldap::Version::Version3);
+				ldap->connect();
+				auto login_details = m_ctx.userLogin();
+				ldap->bindSasl(login_details.user, login_details.password);
+				auto ldap_groups = ldap::getGroupsForUser(ldap, BrokerApp::instance()->cliOptions()->ldapSearchBaseDN(), BrokerApp::instance()->cliOptions()->ldapSearchAttr(), login_details.user);
+				auto group_mapping = BrokerApp::instance()->cliOptions()->ldapGroupMapping();
```
[the remark this answers is not in the document]
Comment (syyyr) [reply libshvbroker/src/brokerapp.cpp:549]: It seems to work fine

[the remark this answers is not in the document]
Comment (syyyr) [reply libshvbroker/src/brokerapp.cpp:549]: CLIOPTION_GETTER_SETTER2 just calls to<RpcValue::List>:
```cpp
	public: ptype getter_prefix##name_rest() const { \
		shv::chainpack::RpcValue val = value(pkey); \
		return val.to<ptype>(); \
	} \
```

File: libshvbroker/src/brokerapp.cpp [hunk 3]
```diff
@@ -517,9 +523,70 @@ bool BrokerApp::checkTunnelSecret(const std::string &s)
...
+	}
+	void run() override
+	{
+		if (auto result = BrokerApp::instance()->aclManager()->checkPassword(m_ctx); result.passwordOk) {
+			emit resultReady(result);
+			return;
+		}
+
+#ifdef WITH_SHV_LDAP
+		if (BrokerApp::instance()->cliOptions()->ldapHostname_isset()) {
```
[the remark this answers is not in the document]
Comment (syyyr) [reply libshvbroker/src/brokerapp.cpp:541]: What about this though:
```
		if (auto result = BrokerApp::instance()->aclManager()->checkPassword(m_ctx); result.passwordOk) {
```
I guess that's not fine either...

[the remark this answers is not in the document]
Comment (syyyr) [reply libshvbroker/src/brokerapp.cpp]: I have changed the other thing at least

File: libshvbroker/src/brokerapp.cpp [hunk 4]
```diff
@@ -680,7 +747,19 @@ chainpack::AccessGrant BrokerApp::accessGrantForRequest(rpc::CommonRpcClientHand
...
-		flatten_user_roles = aclManager()->userFlattenRoles(conn->loggedUserName());
+		if (auto user_def = aclManager()->user(conn->loggedUserName()); user_def.isValid()) {
+			flatten_user_roles = aclManager()->userFlattenRoles(conn->loggedUserName(), user_def.roles);
+		}
+
+#ifdef WITH_SHV_LDAP
+		if (BrokerApp::instance()->cliOptions()->ldapHostname_isset()) {
+			if (auto it = m_ldapUserGroups.find(conn->loggedUserName()); it != m_ldapUserGroups.end()) {
+				std::cerr << "it->second" << " = " << it->second << "\n";
+				flatten_user_roles = aclManager()->userFlattenRoles(conn->loggedUserName(), {it->second});
```
[the remark this answers is not in the document]
Comment (syyyr) [reply libshvbroker/src/brokerapp.cpp:803]: fixed

File: libshvbroker/src/brokerapp.h (hunk 1 above)
Comment (syyyr) [reply libshvbroker/src/brokerapp.h:81]: I'm saving it in BrokerApp which is the only place where it's used afterwards. What benefits would saving it in ClientConnectionOnBroker give? In `accessGrantForRequest`, the normal shv user groups are also fetched by name and from the AclManager, not from the Connection class, so I thought I'd keep it consistent. Although I'm not against making another patch that saves both things in the Connection.

File: libshvbroker/src/brokerapp.cpp [hunk 5]
```diff
@@ -680,7 +747,19 @@ chainpack::AccessGrant BrokerApp::accessGrantForRequest(rpc::CommonRpcClientHand
...
 	else {
-		flatten_user_roles = aclManager()->userFlattenRoles(conn->loggedUserName());
+		if (auto user_def = aclManager()->user(conn->loggedUserName()); user_def.isValid()) {
+			flatten_user_roles = aclManager()->userFlattenRoles(conn->loggedUserName(), user_def.roles);
+		}
+
+#ifdef WITH_SHV_LDAP
+		if (BrokerApp::instance()->cliOptions()->ldapHostname_isset()) {
+			if (auto it = m_ldapUserGroups.find(conn->loggedUserName()); it != m_ldapUserGroups.end()) {
+				std::cerr << "it->second" << " = " << it->second << "\n";
```
[the remark this answers is not in the document]
Comment (syyyr) [reply libshvbroker/src/brokerapp.cpp]: done

**Pull request closed without merging**